library/file.lisp

```diff
@@ -0,0 +1,665 @@
+;;;
+
+(coalton-toplevel
+
+  (define-type FileError
+    "Errors for file functions."
+    (FileError String)
+    (PathError String Pathname)
+    (LispError LispCondition)
+    EOF)
+
+  (define-instance (Signalable FileError)
+    (define (error ferr)
+      (match ferr
+        ((FileError str1)
+         (error (lisp String (str1)
+                  (cl:format cl:nil "File Error:~%~%~a" str1))))
+        ((PathError str path)
+         (error (lisp String (str path)
+                  (cl:format cl:nil "Path Error:~%~%~a ~a" str path))))
+        ((LispError c)
+         (error (lisp String (c)
+                  (cl:format cl:nil "Captured Lisp Error:~%~%~a" c))))
+        ((EOF)
+         (error (lisp String ()
+                  (cl:format cl:nil "Error:~%~%End of File"))))))))
+
+;;;
+;;; Handling existential path queries
+;;;
+
+(coalton-toplevel
+
+  (declare merge ((Into :a Pathname) (Into :b Pathname) => :a -> :b -> (Result FileError Pathname)))
+  (define (merge path1 path2)
+    "Merges two pathnames together. The order is important- for adding subdirectories, make sure to put the subpath (the one without the leading `/`) in `path1`."
+    (let p1 = (the Pathname (into path1)))
+    (let p2 = (the Pathname (into path2)))
+    (pure (lisp Pathname (p1 p2)
+            (cl:merge-pathnames p1 p2))))
+  
+  (declare directory-pathname? ((Into :a Pathname) => :a -> (Result FileError Boolean)))
+  (define (directory-pathname? path)
+    "Returns whether a pathname is a valid directory pathname. Specifically it checks whether it has a trailing `/`."
+    (do
+     (let p = (the Pathname (into path)))
+     (pure (lisp Boolean (p)
+             (to-boolean (uiop:directory-pathname-p p))))))
+
+  (declare file-pathname? ((Into :a Pathname) => :a -> (Result FileError Boolean)))
+  (define (file-pathname? path)
+    "Returns whether a pathname is a valid file pathname."
+    (do
+     (let p = (the Pathname (into path)))
+     (pure (lisp Boolean (p)
+             (to-boolean (uiop:file-pathname-p p))))))
+
+  (declare exists? ((Into :a Pathname) => :a -> (Result FileError Boolean)))
+  (define (exists? path)
+    "Returns whether a file or directory exists."
+    (do
+     (let p = (the Pathname (into path)))
+     (lisp (Result FileError Boolean) (p)
+         (handle-file-function (to-boolean (cl:probe-file p))))))
+
+  (declare if-directory-path ((Into :a Pathname) => :a -> (Pathname -> (Result FileError :b)) -> (Result FileError :b)))
+  (define (if-directory-path path action)
+    "Performs an operation only if the path is a valid directory pathname."
+    (do
+     (let p = (the Pathname (Into path)))
+     (dir <- (directory-pathname? p))
+      (if dir
+          (action p)
+          (Err (PathError "This path does not represent a directory. Please add a trailing `/` or rethink your decision." p)))))
+
+  (declare if-file-path ((Into :a Pathname) => :a -> (Pathname -> (Result FileError :b)) -> (Result FileError :b)))
```

Comment:
Just trying to avoid boilerplate, 7 functions use `if-directory-path`, each saving 5 lines of code.

I can take out `if-file-path` because currently only one function uses it.

I can also make sure to not export them if they're kept in.

Reply:
I removed `if-file-path`, and kept `if-directory-path` as a private function